### ISSUE TITLE
wasm sourcemap not loading

### ISSUE THREAD

**Issue by alexzielenski** [hbenl/vscode-firefox-debug#64]
Using Firefox Quantum 58.0.2

I have a web assembly file being loaded by Firefox (logged as `INFO |000.378|PathConversion: Can't convert url wasm:http://localhost:8000/index.js to path`). The sourcemap is located at `localhost:8000/index.wasm.map` and the wasm at `localhost:8000/index.wasm`. 

Firefox's built-in debugger detects my sourcemap and file tree if I refresh the page with the debugger devtool window open and allows me to set breakpoints.

My issue is that my breakpoints remain unverified and my sources undiscovered while debugging with this extension despite the existence of a valid sourcemap detected by Firefox. I've attempted both `client` and `server` options for the `sourceMap` setting to no avail.

I understand this is a niche use-case for your extension but I wonder if it might be possible for me to use a config to manually specify a souremap to load or if I'm naming my file incorrectly.

Other relevant logs:

```
DEBUG|000.377|DebugConnection: Received response/event {"from":"server1.conn8.child1/tab1","type":"newSource","source":{"actor":"server1.conn8.child1/source31","generatedUrl":null,"url":"wasm:http://localhost:8000/index.js","isBlackBoxed":false,"isPrettyPrinted":false,"isSourceMapped":false,"sourceMapURL":null,"introductionUrl":"wasm:http://localhost:8000/index.js","introductionType":"wasm"}}
DEBUG|000.377|TabActorProxy: Ignored newSource event from tab server1.conn8.child1/tab1
DEBUG|000.377|DebugConnection: Received response/event {"from":"server1.conn8.child1/context25","type":"newSource","source":{"actor":"server1.conn8.child1/source31","generatedUrl":null,"url":"wasm:http://localhost:8000/index.js","isBlackBoxed":false,"isPrettyPrinted":false,"isSourceMapped":false,"sourceMapURL":null,"introductionUrl":"wasm:http://localhost:8000/index.js","introductionType":"wasm"}}
DEBUG|000.377|ThreadActorProxy: New source wasm:http://localhost:8000/index.js on thread server1.conn8.child1/context25
DEBUG|000.377|SourceMappingThreadActorProxy: Trying to sourcemap {"actor":"server1.conn8.child1/source31","generatedUrl":null,"url":"wasm:http://localhost:8000/index.js","isBlackBoxed":false,"isPrettyPrinted":false,"isSourceMapped":false,"sourceMapURL":null,"introductionUrl":"wasm:http://localhost:8000/index.js","introductionType":"wasm"}
```

It seems that the `sourceMapURL` key generated by Firefox is null despite its functionality in the built-in debugger.

Edit: I tried with Firefox Developer Edition 60.0b1 with similar results. Different output, though:
`DEBUG|000.347|PathConversion: Converted url http://localhost:8000/index.wasm to path /Users/Alex/Code/web/wasm/build/debug/index.wasm`

Also, if I use the attach mode of the config, and reload with the debugger tab of the Firefox devtool open, I see all of my sources being converted to proper paths in the console; but breakpoints are failing to be set

```
DEBUG|000.887|BreakpointsAdapter: 0 breakpoints were previously set for http://localhost:8000/main.cpp
DEBUG|000.888|SourceActorProxy: Setting breakpoint at line 14 and column undefined in http://localhost:8000/main.cpp
DEBUG|000.888|DebugConnection: Sending request {"to":"server1.conn4.child1/source130","type":"setBreakpoint","location":{"line":14}}
DEBUG|000.888|BreakpointsAdapter: Adding 1 and removing 0 breakpoints
DEBUG|000.889|ThreadCoordinator: state: paused, target: running, tasks: 1/0, eval: 0
DEBUG|003.087|DebugConnection: Received response/event {"actor":"server1.conn4.child1/breakpoint132","isPending":true,"from":"server1.conn4.child1/source130"}
DEBUG|003.087|SourceActorProxy: **Breakpoint has been set at undefined** in http://localhost:8000/main.cpp
DEBUG|003.088|BreakpointsAdapter: Updating breakpoints
DEBUG|003.088|ThreadCoordinator: state: paused, target: running, tasks: 0/0, eval: 0
DEBUG|003.089|ThreadPauseCoordinator: Requesting resume for Tab 1
```

**Comment by hbenl** [comment]
I've never worked with wasm, so this use case was completely untested - so far. I'd like to support this, could you point me to a good introductory wasm example project?

**Comment by alexzielenski** [comment]
[wasmExample.zip](https://github.com/hbenl/vscode-firefox-debug/files/1794358/wasmExample.zip)

I've attached a zip file of an example project I just threw together to demonstrate this. It's preconfigured for use with VS code and your extension on attach mode. 

Interestingly enough, when I tested it, the browser successfully stopped at my breakpoint! Something else must be stopping my more complicated project from resolving breakpoint locations.

To use the project you need to start a local web server on port 8000 in the build/debug folder (I've included a prebuilt copy) and browse to it in Firefox Developer Edition. Next use VS Code to attach its debugger to Firefox.

If you would like to be able to build wasm from the cpp source, you need only install emscripten. (`brew install emscripten` on macOS) and run `make debug`.

Note: the include paths are configured for OS X 10.13, you'll need to adjust them accordingly.

**Comment by hbenl** [comment]
Thanks for the example, but I can't find any `.wasm` file in there, only `index.js`, which looks like asm.js format to me - it seems that the `-s WASM=1` parameter in the `Makefile` is somehow ignored. Could you please update the example to compile to wasm?

Also, I've noticed that the sourcemap file `build/debug/index.js.map` contains the reference `src/main.cpp`, which means that my adapter will try to map to a file `build/debug/src/main.cpp`, which doesn't exist - that could be the reason why you can't set breakpoints.
This should be fixed by adding a base url to the sourcemap, which I would expect the `--source-map-base` parameter in the `Makefile` should do, but somehow it doesn't add the base url.
It could also be fixed by adding a corresponding path mapping to `launch.json`. However, my adapter currently ignores `pathMappings` for sourcemapped files when `sourceMaps` is set to `client` - I will fix this shortly. Perhaps this will make breakpoints work in your project.

**Comment by hbenl** [comment]
Correction: `pathMappings` *are* applied even if `sourceMaps` is set to `client` - so that's not the issue here...

On another note, when you include log messages, please always state the `sourceMaps` setting that the logs were captured with - it makes a big difference when trying to interpret the logs.

**Comment by alexzielenski** [comment]
Thanks for taking a look. You are right in that it did not compile wasm due to a typo in the makefile. I had figured since I set `pathMappings` in my `launch.json` that the paths would resolve and my breakpoints do end up being verified this way. 

Here is a copy with the fixed makefile and wasm output. I also added to it a command so that it copies source files on build to `buid/debug`.

[wasmExample.zip](https://github.com/hbenl/vscode-firefox-debug/files/1798805/wasmExample.zip)

**Comment by hbenl** [comment]
This zip file contains the same files as before and no `.wasm` file.

**Comment by alexzielenski** [comment]
[wasmExample.zip](https://github.com/hbenl/vscode-firefox-debug/files/1799597/wasmExample.zip)

Third time's the charm. (Uploaded the old file last time, oops). I double checked it. Thanks for your patience.

**Comment by hbenl** [comment]
Thanks, I could reproduce it now and added experimental support for WebAssembly debugging.

Note that WebAssembly debugging only works when Firefox' built-in debugger is opened (even if you use VS Code for debugging). I found [this issue](https://github.com/devtools-html/debugger.html/issues/4547) which sounds like this is the desired behaviour.

Furthermore, you will have to set your breakpoints in the copied file under `build/debug/src`, otherwise it won't work. Unfortunately, `pathMappings` don't work properly in this case and I'm not yet sure how to fix this without breaking them in other situations.

*And* you have to set `sourceMaps` to `client`.

I'd love to hear back from you if it works in your project. To be honest, in this case I don't really know what I'm doing, since the changes Mozilla made to the debugger protocol are undocumented (and some of them are pretty weird). I'll have to reverse engineer this by looking at the source of Firefox' built-in debugger. That may take some time...

**Comment by alexzielenski** [comment]
Hey! Thanks for the update! It works! My web assembly successfully stopped at the correct point the highlighted the correct line. I attached a screenshot. And for my use case, it didn't actually seem necessary to copy the folders into build/debug. The `pathMappings` seem to be working for verifying the breakpoints atleast with my `sourceMaps` set to `server` despite your suggestion.

<img width="925" alt="screen shot 2018-03-11 at 5 13 22 pm" src="https://user-images.githubusercontent.com/351783/37258518-ba2ce110-254f-11e8-9b46-cfa604ea5855.png">

**Comment by hbenl** [comment]
Thanks for the feedback!
I will fix those error messages from the `FrameAdapter` soon and I also have an idea how to fix the `pathMappings` for the client-side sourcemaps now. It's interesting that server-side sourcemaps work for you, they still don't worked for me.

**Comment by hbenl** [comment]
both the `FrameAdapter` error messages and the `pathMappings` for the client-side sourcemaps are fixed in version 1.1.3
